src/Http/ServerErrorHandler.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the ACME PHP library.
+ *
+ * (c) Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace AcmePhp\Core\Http;
+
+use AcmePhp\Core\Exception\AcmeCoreServerException;
+use GuzzleHttp\Exception\RequestException;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Create appropriate exception for given server response.
+ *
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class ServerErrorHandler
+{
+    private static $exceptions = [
+        'badCSR'         => 'BadCsrServerException',
+        'badNonce'       => 'BadNonceServerException',
+        'connection'     => 'ConnectionServerException',
+        'serverInternal' => 'InternalServerException',
+        'invalidEmail'   => 'InvalidEmailServerException',
+        'malformed'      => 'MalformedServerException',
+        'rateLimited'    => 'RateLimitedServerException',
+        'tls'            => 'TlsServerException',
+        'unauthorized'   => 'UnauthorizedServerException',
+        'unknownHost'    => 'UnknownHostServerException',
+    ];
+
+    /**
+     * @param RequestInterface  $request
+     * @param ResponseInterface $response
+     * @param \Exception|null   $previous
+     *
+     * @return AcmeCoreServerException
+     */
+    public function createAcmeExceptionForResponse(
+        RequestInterface $request,
+        ResponseInterface $response,
+        \Exception $previous = null
+    ) {
+        $body = \GuzzleHttp\Psr7\copy_to_string($response->getBody());
+        $data = @json_decode($body, true);
+
+        if (!$data || !isset($data['type'], $data['detail'])) {
+            // Not JSON: not an ACME error response
+            return $this->createDefaultExceptionForResponse($request, $response, $previous);
+        }
+
+        // Remove "urn:acme:error:" prefix
+        $type = substr($data['type'], 15);
```

Comment:
prefere a regular expression or asserting that the 15 first chart are really `urn:acme:error`